### PR TITLE
Improve varchar support for hive connector

### PR DESCRIPTION
**Pull request by arhimondr**

### CODE REVIEW

**Comment by losipiuk** [comment]
Add commit messages explaining behaviour semantics.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveType.java [hunk 1]
```diff
@@ -225,7 +228,14 @@ private static TypeInfo toTypeInfo(Type type)
...
         if (type instanceof VarcharType) {
-            return HIVE_STRING.typeInfo;
+            VarcharType varcharType = (VarcharType) type;
+            int varcharLength = varcharType.getLength();
+            if (varcharLength <= MAX_VARCHAR_LENGTH) {
+                return getVarcharTypeInfo(varcharLength);
+            }
+            else if (varcharLength == MAX_LENGTH) {
+                return HIVE_STRING.typeInfo;
+            }
```
Comment (losipiuk) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveType.java]: It would be nice to throw specific exception here so user knows what values for precision parameter are supported.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveWriteUtils.java [hunk 2]
```diff
@@ -504,8 +508,15 @@ public static ObjectInspector getRowColumnInspector(Type type)
...
 
-        if (type.equals(VarcharType.VARCHAR)) {
-            return writableStringObjectInspector;
+        if (type instanceof VarcharType) {
+            VarcharType varcharType = (VarcharType) type;
+            int varcharLength = varcharType.getLength();
+            if (varcharLength <= MAX_VARCHAR_LENGTH) {
+                return getPrimitiveWritableObjectInspector(getVarcharTypeInfo(varcharLength));
+            }
+            else if (varcharLength == MAX_LENGTH) {
```
Comment (losipiuk) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveWriteUtils.java]: do not static import this

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveWriteUtils.java [hunk 3]
```diff
@@ -504,8 +508,15 @@ public static ObjectInspector getRowColumnInspector(Type type)
             return writableDoubleObjectInspector;
         }
 
-        if (type.equals(VarcharType.VARCHAR)) {
-            return writableStringObjectInspector;
+        if (type instanceof VarcharType) {
+            VarcharType varcharType = (VarcharType) type;
+            int varcharLength = varcharType.getLength();
+            if (varcharLength <= MAX_VARCHAR_LENGTH) {
```
Comment (losipiuk) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveWriteUtils.java]: do not static import this.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveWriteUtils.java [hunk 4]
```diff
@@ -504,8 +508,15 @@ public static ObjectInspector getRowColumnInspector(Type type)
             return writableDoubleObjectInspector;
         }
 
-        if (type.equals(VarcharType.VARCHAR)) {
-            return writableStringObjectInspector;
+        if (type instanceof VarcharType) {
+            VarcharType varcharType = (VarcharType) type;
+            int varcharLength = varcharType.getLength();
```
Comment (losipiuk) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveWriteUtils.java:512]: add comment what is happening here

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java [hunk 5]
```diff
@@ -367,97 +354,91 @@ public static boolean isHiveNull(byte[] bytes)
...
     }
 
-    public static NullableValue parsePartitionValue(String partitionName, String value, HiveType hiveType, DateTimeZone timeZone)
+    public static NullableValue parsePartitionValue(String partitionName, String value, Type type, DateTimeZone timeZone)
     {
         boolean isNull = HIVE_DEFAULT_DYNAMIC_PARTITION.equals(value);
 
-        Optional<DecimalType> decimalType = getDecimalType(hiveType);
-        if (decimalType.isPresent()) {
+        if (type instanceof DecimalType) {
```
Comment (losipiuk) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java:361]: split refactoring and actual changes to separate commits

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 6]
```diff
@@ -751,8 +754,8 @@ public void testShowColumnsPartitionKey()
     {
         assertUpdate("" +
                 "CREATE TABLE test_show_columns_partition_key\n" +
-                "(grape bigint, orange bigint, pear varchar(65535), mango integer, apple varchar)\n" +
-                "WITH (partitioned_by = ARRAY['apple'])");
+                "(grape bigint, orange bigint, pear varchar(65535), mango integer, apple varchar, lime varchar(65535))\n" +
```
Comment (losipiuk) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: use ananas instead of lime for consistency.
Other fruits are sweet ones.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java [hunk 7]
```diff
@@ -367,97 +354,91 @@ public static boolean isHiveNull(byte[] bytes)
...
-            if (HIVE_TIMESTAMP.equals(hiveType)) {
-                if (isNull) {
-                    return NullableValue.asNull(TIMESTAMP);
-                }
-                return NullableValue.of(TIMESTAMP, parseHiveTimestamp(value, timeZone));
+        if (DATE.equals(type)) {
+            if (isNull) {
+                return NullableValue.asNull(DATE);
             }
+            return NullableValue.of(DATE, datePartitionKey(value, partitionName));
```
Comment (losipiuk) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java:414]: were all those `*partitioningKey` methods already defined? Why were those not used before?

**Comment by losipiuk** [comment]
Looks good modulo comments

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java (hunk 5 above)
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java:361]: I can't split it. There are no changes left after refactoring.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java (hunk 6 above)
Comment (arhimondr) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: Sure, i haven't thought about that =)

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java (hunk 7 above)
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java:414]: Because we hadn't pass the presto Type parameter. So we had to duplicate all the parsing twice. Now  instead of passing `HiveType` i just pass the Presto `Type`. So i can reuse the partition key parsing methods.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveWriteUtils.java (hunk 4 above)
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveWriteUtils.java:512]: done
